lib/spack/spack/util/windows_registry.py

```diff
@@ -0,0 +1,313 @@
+        return True
+
+    @property
+    def reg(self):
+        if not self._reg:
+            self._load_key()
+        return self._reg
+
+    def get_value(self, value_name):
+        """Return registry value corresponding to provided argument (if it exists)"""
+        if not self._valid_reg_check():
+            raise RegistryError("Cannot query value from invalid key %s" % self.key)
+        with self.invalid_reg_ref_error_handler():
+            return self.reg.get_value(value_name)
+
+    def get_subkey(self, subkey_name):
+        if not self._valid_reg_check():
+            raise RegistryError("Cannot query subkey from invalid key %s" % self.key)
+        with self.invalid_reg_ref_error_handler():
+            return self.reg.get_subkey(subkey_name)
+
+    def get_subkeys(self):
+        if not self._valid_reg_check():
+            raise RegistryError("Cannot query subkeys from invalid key %s" % self.key)
+        with self.invalid_reg_ref_error_handler():
+            return self.reg.subkeys
+
+    def get_values(self):
+        if not self._valid_reg_check():
+            raise RegistryError("Cannot query values from invalid key %s" % self.key)
+        with self.invalid_reg_ref_error_handler():
+            return self.reg.values
+
+    def _traverse_subkeys(self, stop_condition):
+        """Perform simple BFS of subkeys, returning the key
+        that successfully triggers the stop condition.
+        Args:
+            stop_condition: lambda or function pointer that takes a single argument
+                            a key and returns a boolean value based on that key
+        Return:
+            the key if stop_condition is triggered, or None if not
+        """
+        if not self._valid_reg_check():
+            return None
+        with self.invalid_reg_ref_error_handler():
+            queue = self.reg.subkeys
+            for key in queue:
+                if stop_condition(key):
+                    return key
+                queue.extend(key.subkeys)
+            return None
+
+    def find_subkey(self, subkey_name, recursive=True):
+        """If non recursive, this method is the same as get subkey with error handling
+        Otherwise perform a BFS of subkeys until desired key is found
+        Returns None or RegistryKey object corresponding to requested key name
+
+        Args:
+            subkey_name (str): string representing subkey to be searched for
+            recursive (bool): optional argument, if True, subkey need not be a direct
+                                sub key of this registry entry, and this method will
+                                search all subkeys recursively.
+                                Default is True
+        Return:
+            the desired subkey as a RegistryKey object, or none
+        """
+        if not recursive:
+            return self.get_subkey(subkey_name)
+
+        else:
+            return self._traverse_subkeys(lambda x: x.name == subkey_name)
```

Comment:
I think this and `find_value` should avoid calling `_traverse_subkeys` if `not self._valid_reg_check()`, and `_traverse_subkeys` should fail if the root key does not exist (IMO allowing traversals starting from an invalid root is confusing). Would that cause problems?

Reply:
I had intended to just consolidate the error handling for both into `_traverse_subkeys`, I just missed updating traverse subkeys to throw like the other methods do now.